drivers/opamp/opamp_stm32g4x_opamp.c

```diff
@@ -0,0 +1,666 @@
+	const struct adc_dt_spec adc_ch; /* ADC channel opamp is connected to */
+	bool lock_enable;
+	bool self_calibration;
+	uint32_t inputs_mux_mode;
+	opamp_inm_filtering_t inm_filtering;
+	uint8_t pmos_trimming_value;
+	uint8_t nmos_trimming_value;
+	const struct stm32_pclken *pclken;
+	const size_t pclk_len;
+	const struct pinctrl_dev_config *pincfg;
+};
+
+struct stm32_opamp_data {
+	struct k_mutex dev_mtx; /* Sync between device accesses */
+};
+
+static inline void stm32_opamp_config_log_dbg(const struct device *dev)
+{
+	const struct stm32_opamp_config *cfg = dev->config;
+
+	LOG_DBG("%s config:\n"
+		"  functional_mode: 0x%x\n"
+		"  power_mode: 0x%x\n"
+		"  inm: {0x%08x,0x%08x}\n"
+		"  inp: {0x%08x,0x%08x}\n"
+		"  adc_ch: %s channel: 0x%x\n"
+		"  lock_enable: %s\n"
+		"  self_calibration: %s\n"
+		"  inputs_mux_mode: 0x%x\n"
+		"  inm_filtering: 0x%x\n"
+		"  pmos_trimming_value: 0x%02x\n"
+		"  nmos_trimming_value: 0x%02x\n"
+		"  OPAMPx_CSR: 0x%08x\n"
+		"  OPAMPx_TCMR: 0x%08x\n",
+		dev->name, cfg->functional_mode, cfg->power_mode, cfg->inm[0], cfg->inm[1],
+		cfg->inp[0], cfg->inp[1], cfg->adc_ch.dev ? cfg->adc_ch.dev->name : "none",
+		cfg->adc_ch.channel_id, cfg->lock_enable ? "true" : "false",
+		cfg->self_calibration ? "true" : "false", cfg->inputs_mux_mode, cfg->inm_filtering,
+		(uint32_t)cfg->pmos_trimming_value, (uint32_t)cfg->nmos_trimming_value,
+		(uint32_t)cfg->opamp->CSR, (uint32_t)cfg->opamp->TCMR);
+}
+
+static bool stm32_opamp_is_locked(const struct device *dev)
+{
+	const struct stm32_opamp_config *cfg = dev->config;
+	OPAMP_TypeDef *opamp = cfg->opamp;
+
+	if (cfg->inputs_mux_mode != LL_OPAMP_INPUT_MUX_DISABLE) {
+		return LL_OPAMP_IsLocked(opamp) && LL_OPAMP_IsTimerMuxLocked(opamp);
+	}
+
+	return LL_OPAMP_IsLocked(opamp);
+}
+
+static void stm32_opamp_lock(const struct device *dev)
+{
+	const struct stm32_opamp_config *cfg = dev->config;
+	OPAMP_TypeDef *opamp = cfg->opamp;
+
+	LL_OPAMP_Lock(opamp);
+	if (cfg->inputs_mux_mode != LL_OPAMP_INPUT_MUX_DISABLE) {
+		LL_OPAMP_LockTimerMux(opamp);
+	}
+}
+
+static int stm32_opamp_pm_callback(const struct device *dev, enum pm_device_action action)
+{
+	const struct stm32_opamp_config *cfg = dev->config;
+	OPAMP_TypeDef *opamp = cfg->opamp;
+
+	if (stm32_opamp_is_locked(dev)) {
+		return -EPERM;
+	}
+
+	if (action == PM_DEVICE_ACTION_RESUME) {
+		LL_OPAMP_Enable(opamp);
+		if (cfg->lock_enable) {
+			stm32_opamp_lock(dev);
+		}
+	}
+
+	if (action == PM_DEVICE_ACTION_SUSPEND) {
+		LL_OPAMP_Disable(opamp);
+	}
+
+	return 0;
+}
+
+static int stm32_opamp_set_gain(const struct device *dev, enum opamp_gain gain)
+{
+	const struct stm32_opamp_config *cfg = dev->config;
+	struct stm32_opamp_data *data = dev->data;
+
+	k_mutex_lock(&data->dev_mtx, K_FOREVER);
+
+	OPAMP_TypeDef *opamp = cfg->opamp;
+	uint32_t stm32_ll_gain = UINT32_MAX;
+
+	if (stm32_opamp_is_locked(dev)) {
+		LOG_ERR("%s: locked", dev->name);
+		k_mutex_unlock(&data->dev_mtx);
+		return -EPERM;
```

Comment:
Could this return with success if the requested gain matches the already programmed one?
Discard this comment if you think the scenario is not applicable.

Reply:
Sure, it makes sense. Let return OK (e.g. 0) if gain matches the old one. I think I could move lock chick rignt before LL_SetPGA() function call and before lock check PGA gain could be compared with LL PGA gain determined in if-else statement. Well, take a look on it as sonn as I request your review again.